// yandex-mobile-ads/lib/module/MobileAds.js
import YandexMobileAdsModule from './YandexMobileAdsModule';
import { isBoolean, isSupportedPlatform } from './util';
class MobileAdsModule {
  pluginVersion = '7.13.0';
  getLibraryVersion() {
    try {
      isSupportedPlatform();
      return YandexMobileAdsModule.YandexMobileAds.getLibraryVersion();
    } catch {}
  }
  initialize() {
    try {
      isSupportedPlatform();
      return YandexMobileAdsModule.YandexMobileAds.initializeSdk();
    } catch {}
  }
  showDebugPanel() {
    try {
      isSupportedPlatform();
      return YandexMobileAdsModule.YandexMobileAds.showDebugPanel();
    } catch {}
  }
  enableLogging(enableLogging) {
    try {
      isSupportedPlatform();
      isBoolean(enableLogging);
      YandexMobileAdsModule.YandexMobileAds.enableLogging(enableLogging);
    } catch {}
  }
  enableDebugErrorIndicator(enableIndicator) {
    try {
      isSupportedPlatform();
      isBoolean(enableIndicator);
      YandexMobileAdsModule.YandexMobileAds.enableDebugErrorIndicator(enableIndicator);
    } catch {}
  }
  setLocationConsent(locationConsent) {
    try {
      isSupportedPlatform();
      isBoolean(locationConsent);
      YandexMobileAdsModule.YandexMobileAds.setLocationConsent(locationConsent);
    } catch {}
  }
  setAgeRestrictedUser(ageRestrictedUser) {
    try {
      isSupportedPlatform();
      isBoolean(ageRestrictedUser);
      YandexMobileAdsModule.YandexMobileAds.setAgeRestrictedUser(ageRestrictedUser);
    } catch {}
  }
  setUserConsent(userConsent) {
    try {
      isSupportedPlatform();
      isBoolean(userConsent);
      YandexMobileAdsModule.YandexMobileAds.setUserConsent(userConsent);
    } catch {}
  }
}
const MobileAdsInstance = new MobileAdsModule();
export const MobileAds = MobileAdsInstance;
export default MobileAds;
